File: week14/willow2150/16198.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws Exception {
		BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(reader.readLine());
		StringTokenizer tokenizer = new StringTokenizer(reader.readLine());

		int[] list = new int[N];
		for (int i = 0; i < N; i++)
			list[i] = Integer.parseInt(tokenizer.nextToken());

		System.out.print(recursion(list, N - 1, N));
	}

	public static int recursion(int[] list, int lastIndex, int size) {
		if (size == 2)
			return 0;
		size--;

		int maxValue = 0;

		for (int index = 1; index < lastIndex; index++) {
			if (list[index] == 0)
				continue;
			int leftIndex = index - 1;
			int rightIndex = index + 1;
			while (0 < leftIndex && list[leftIndex] == 0)
				leftIndex--;
			while (rightIndex < lastIndex && list[rightIndex] == 0)
				rightIndex++;
			int temp = list[index];
			list[index] = 0;
			maxValue = Math.max(
				maxValue,
				list[leftIndex] * list[rightIndex] + recursion(list, lastIndex, size)
			);
			list[index] = temp;
		}
		return maxValue;
	}
}

File: week14/willow2150/16198.py
from sys import stdin


def recursion(size):
    if size == 2:
        return 0
    size -= 1
    max_value = 0

    for index in range(1, last_index):
        if weight_list[index] == 0:
            continue
        left_index = index - 1
        right_index = index + 1
        while 0 < left_index and weight_list[left_index] == 0:
            left_index -= 1
        while right_index < last_index and weight_list[right_index] == 0:
            right_index += 1
        temp = weight_list[index]
        weight_list[index] = 0
        max_value = max(
            max_value,
            weight_list[left_index] * weight_list[right_index] + recursion(size)
        )
        weight_list[index] = temp
    return max_value


N = int(stdin.readline())
weight_list = list(map(int, stdin.readline().split()))
last_index = len(weight_list) - 1
print(recursion(len(weight_list)))
